File: querier/internal/paths/paths_test.go
package paths

import (
	"path/filepath"
	"testing"
)

func TestResolveFor(t *testing.T) {
	tests := []struct {
		name     string
		explicit string
		env      map[string]string
		goos     string
		userHome string
		want     string
	}{
		{
			name:     "explicit",
			explicit: "/custom",
			env:      map[string]string{"RESEARCH4JAR_HOME": "/environment"},
			goos:     "linux",
			userHome: "/home/test",
			want:     "/custom",
		},
		{
			name:     "environment",
			env:      map[string]string{"RESEARCH4JAR_HOME": "/environment"},
			goos:     "darwin",
			userHome: "/Users/test",
			want:     "/environment",
		},
		{
			name:     "macOS",
			env:      map[string]string{},
			goos:     "darwin",
			userHome: "/Users/test",
			want:     "/Users/test/Library/Application Support/research4jar",
		},
		{
			name:     "Linux XDG",
			env:      map[string]string{"XDG_DATA_HOME": "/xdg"},
			goos:     "linux",
			userHome: "/home/test",
			want:     "/xdg/research4jar",
		},
		{
			name:     "Linux fallback",
			env:      map[string]string{},
			goos:     "linux",
			userHome: "/home/test",
			want:     "/home/test/.local/share/research4jar",
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			got, err := ResolveFor(test.explicit, test.env, test.goos, test.userHome)
			if err != nil {
				t.Fatal(err)
			}
			want, err := filepath.Abs(filepath.FromSlash(test.want))
			if err != nil {
				t.Fatal(err)
			}
			if got.Home != want {
				t.Fatalf("Home = %q, want %q", got.Home, want)
			}
			if got.Manifest != filepath.Join(got.Home, "manifest.db") {
				t.Fatalf("Manifest = %q", got.Manifest)
			}
		})
	}
}

func TestResolveForWindowsRequiresLocalAppData(t *testing.T) {
	_, err := ResolveFor("", map[string]string{}, "windows", `C:\Users\test`)
	if err == nil {
		t.Fatal("expected an error")
	}
}

File: core/src/test/kotlin/dev/research4jar/indexer/PathsTest.kt
package dev.research4jar.indexer

import java.nio.file.Path
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class PathsTest {
    @Test
    fun `explicit home wins`() {
        val paths = Research4JarPaths.resolve(
            explicitHome = "/custom/research4jar",
            environment = mapOf("RESEARCH4JAR_HOME" to "/environment"),
            osName = "Linux",
            userHome = "/home/user",
        )
        assertEquals(expected("/custom/research4jar"), paths.home.toString())
    }

    @Test
    fun `environment home wins over platform default`() {
        val paths = Research4JarPaths.resolve(
            environment = mapOf("RESEARCH4JAR_HOME" to "/environment"),
            osName = "Mac OS X",
            userHome = "/Users/test",
        )
        assertEquals(expected("/environment"), paths.home.toString())
    }

    @Test
    fun `platform defaults follow specification`() {
        assertEquals(
            expected("/Users/test/Library/Application Support/research4jar"),
            Research4JarPaths.resolve(
                environment = emptyMap(),
                osName = "Mac OS X",
                userHome = "/Users/test",
            ).home.toString(),
        )
        assertEquals(
            expected("/xdg/research4jar"),
            Research4JarPaths.resolve(
                environment = mapOf("XDG_DATA_HOME" to "/xdg"),
                osName = "Linux",
                userHome = "/home/test",
            ).home.toString(),
        )
        assertEquals(
            expected("/home/test/.local/share/research4jar"),
            Research4JarPaths.resolve(
                environment = emptyMap(),
                osName = "Linux",
                userHome = "/home/test",
            ).home.toString(),
        )
    }

    @Test
    fun `windows requires local app data`() {
        assertFailsWith<IllegalStateException> {
            Research4JarPaths.resolve(
                environment = emptyMap(),
                osName = "Windows 11",
                userHome = "C:\\Users\\test",
            )
        }
    }

    @Test
    fun `darwin is not mistaken for windows`() {
        val paths = Research4JarPaths.resolve(
            environment = emptyMap(),
            osName = "Darwin",
            userHome = "/Users/test",
        )
        assertEquals(
            expected("/Users/test/Library/Application Support/research4jar"),
            paths.home.toString(),
        )
    }

    private fun expected(value: String): String =
        Path.of(value).toAbsolutePath().normalize().toString()
}
